File: python/01-getting-started/ex01_camera_stream_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
@file ex01_camera_stream_test.py
@brief Exercise 10: Real-Time Stream Ingestion and Serialized Output Pipeline (`cv2.VideoCapture` + `cv2.VideoWriter`)

@version 1.0.0
@date 2026-08-16
@author Matha Goram (Systems Architect)

@copyright MIT License
Copyright (c) 2026 ParkCircus Productions. All Rights Reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

@par Purpose
Demonstrates establishing an automated loop to ingest multiple sequential frames from a live
video stream and write them out directly into a persistent video file container.

@par Update History
- v1.0.0 (2026-08-16): Initial Python port for Chapter 01 getting started.

@par Processing Workflow and Algorithms
1. Initialize `cv2.VideoCapture` and query stream resolution parameters.
2. Setup `cv2.VideoWriter` matching target frame dimensions.
3. Process a bounded batch of sequential frames in an execution loop.
4. Release device handlers and close file streams cleanly.
"""

import sys
import numpy as np
import cv2


def main():
    ex_prefix = "Ex 01 - Camera Stream"
    output_filename = "output_live_stream_record.mp4"

    print(f"[{ex_prefix}] 📹 Initializing live stream record pipeline...")

    cap = cv2.VideoCapture(0)
    use_live_camera = cap.isOpened()

    width, height, fps = 640, 480, 25.0
    if use_live_camera:
        width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH) or 640)
        height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT) or 480)

    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    writer = cv2.VideoWriter(output_filename, fourcc, fps, (width, height))

    if not writer.isOpened():
        print(f"[{ex_prefix}] ❌ ERROR: Could not open VideoWriter target {output_filename}.", file=sys.stderr)
        if use_live_camera:
            cap.release()
        sys.exit(1)

    print(f"[{ex_prefix}] 🔄 Recording stream sequence (30 frames)...")
    for i in range(30):
        if use_live_camera:
            ret, frame = cap.read()
            if not ret or frame is None:
                break
        else:
            # Fallback synthetic frame sequence for headless testing
            frame = np.full((height, width, 3), (30, 30, 30), dtype=np.uint8)
            cv2.putText(frame, f"Simulated Live Frame {i}", (50, 240),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 0), 2)

        writer.write(frame)

    if use_live_camera:
        cap.release()
    writer.release()

    print(f"[{ex_prefix}] ✅ Camera stream recording completed successfully. Saved to {output_filename}.")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: cpp/01-getting-started/ex01_camera_stream_test.cpp
/**
 * @file ex01_camera_stream_test.cpp
 * @brief Test 09: Live Hardware Video Stream Ingestion and Encoding
 *
 * @par Purpose
 * Demonstrates continuous frame-by-frame processing from a live hardware camera via the
 * native GStreamer pipeline (`libcamerasrc`). Establishes a sensor connection, ingests a
 * sequence of real-time frames, applies a dynamic text overlay (HUD) to each matrix buffer,
 * and serializes the sequence into an encoded MP4 video container.
 *
 * @author Matha Goram
 * @version 1.1.0
 * @date 2026-08-12
 *
 * @par Copyright
 * Copyright (c) 2026 ParkCircus Productions. All rights reserved.
 *
 * @par MIT License
 * [Standard MIT License text omitted for brevity]
 *
 * @par Prerequisites
 * - C++17 compliant compiler (GCC / Clang)
 * - CMake 3.22+
 * - OpenCV 4.x or 5.x development libraries with GStreamer support enabled
 * - Active libcamera stack (standard on RPi Bullseye/Bookworm/Trixie)
 * - Required GStreamer Runtime and Plugin Packages:
 *   - `gstreamer1.0-tools`
 *   - `gstreamer1.0-plugins-base`
 *   - `gstreamer1.0-plugins-good`
 *   - `gstreamer1.0-plugins-bad`
 *   - `gstreamer1.0-libcamera`
 *
 * @par User Interface Guide
 * Non-interactive Command-Line Interface (CLI):
 * 1. Build the CMake target `ex01_camera_stream_test`.
 * 2. Execute on the remote node (No wrappers required).
 * 3. Verify generated container `output_live_stream.mp4` via IDE or remote media tools.
 *
 * @par Processing Workflow and Algorithms
 * 1. **Pipeline Initialization**: Open `cv::VideoCapture` passing a GStreamer pipeline string targeting `libcamerasrc` with explicit resolution and framerate settings.
 * 2. **Sink Allocation**: Initialize `cv::VideoWriter` targeting an MP4 container matching the stream dimensions.
 * 3. **Stream Loop**: Execute a bounded loop for $N$ frames (e.g., 90 frames for a 3-second recording burst).
 * 4. **Buffer Processing**: For every iteration, extract a frame via `cap.read()`, stamp it with the current frame count via `cv::putText()`, and pass it to the encoder.
 * 5. **Resource Teardown**: Safely release both the camera capture and video writer handles to prevent memory leaks and hardware locks.
 *
 * @par Error Message Responses
 * - `[Ex 01 - Camera Stream] ❌ ERROR: Could not open the GStreamer pipeline.`
 *   - **Cause**: OpenCV built without GStreamer support or missing plugin packages.
 * - `[Ex 01 - Camera Stream] ❌ ERROR: Frame stream dropped prematurely.`
 *   - **Cause**: Pipeline capability mismatch or hardware drop mid-loop.
 *
 * @par References and Notes
 * - **Resolution Steps for Raspberry Pi CSI Cameras (IMX477 / Unicam)**:
 *   1. Avoid standard V4L2 device index bindings (`cv::CAP_V4L2`) on CSI sensors because they output raw Bayer data directly from Unicam, causing `select() timeout` errors.
 *   2. Install the required GStreamer runtime toolchains and plugins: `sudo apt install -y gstreamer1.0-tools gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-libcamera`.
 *   3. Route video capture through a native GStreamer pipeline string via `cv::CAP_GSTREAMER`: `"libcamerasrc ! video/x-raw, width=1280, height=720, framerate=30/1 ! videoconvert ! appsink"`.
 * - **Common Fourcc Options for MP4 Containers**:
 *   - `cv::VideoWriter::fourcc('m', 'p', '4', 'v')`: MPEG-4 Visual codec. Widely supported across platforms with minimal external dependencies.
 *   - `cv::VideoWriter::fourcc('a', 'v', 'c', '1')` or `cv::VideoWriter::fourcc('H', '2', '6', '4')`: H.264 AVC codec. Delivers superior compression and smaller file sizes, though it relies on your system's FFmpeg installation having H.264 encoding support enabled.
 */

#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>
#include <cstdlib>

int main() {
    const std::string ex_prefix = "Ex 01 - Camera Stream";
    const std::string output_filename = "output_live_stream.mp4";

    // Define the GStreamer pipeline targeting the native libcamera stack
    const std::string pipeline = "libcamerasrc ! video/x-raw, width=1280, height=720, framerate=30/1 ! videoconvert ! appsink";

    std::cout << "[" << ex_prefix << "] 🎬 Initializing live sensor telemetry loop via GStreamer (libcamerasrc)...\n";

    // Open the pipeline using the GStreamer backend
    cv::VideoCapture cap(pipeline, cv::CAP_GSTREAMER);
    if (!cap.isOpened()) {
        std::cerr << "[" << ex_prefix << "] ❌ ERROR: Could not open the GStreamer pipeline.\n";
        std::cerr << "          Ensure OpenCV was built with GStreamer support and libcamerasrc is available.\n";
        return EXIT_FAILURE;
    }

    int frame_width = 1280;
    int frame_height = 720;
    int fps = 30;
    int frames_to_record = 90; // 3-second recording burst

    cv::VideoWriter writer(output_filename, cv::VideoWriter::fourcc('m', 'p', '4', 'v'), fps, cv::Size(frame_width, frame_height));

    if (!writer.isOpened()) {
        std::cerr << "[" << ex_prefix << "] ❌ ERROR: Could not open the output video file for write.\n";
        cap.release();
        return EXIT_FAILURE;
    }

    cv::Mat frame;
    std::cout << "[" << ex_prefix << "] ⏳ Recording live stream burst...\n";

    for (int i = 0; i < frames_to_record; i++) {
        if (!cap.read(frame) || frame.empty()) {
            std::cerr << "[" << ex_prefix << "] ❌ ERROR: Frame stream dropped prematurely at frame " << i << ".\n";
            break;
        }

        // Draw a diagnostic Heads-Up Display (HUD) overlay
        std::string hud_text = "LIVE STREAM | Frame: " + std::to_string(i);
        cv::putText(frame, hud_text, cv::Point(20, 40),
                    cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 0), 2, cv::LINE_AA);

        writer.write(frame); // Push modified frame to encoder
    }

    // Safe teardown of all hardware handles
    cap.release();
    writer.release();

    std::cout << "[" << ex_prefix << "] ✅ Hardware stream successfully encoded to: " << output_filename << "\n";
    return EXIT_SUCCESS;
}
